File: blue-core/src/blue/noteProcessor/SwitchProcessor.java
package blue.noteProcessor;

import blue.BlueSystem;
import blue.plugin.NoteProcessorPlugin;
import blue.soundObject.Note;
import blue.soundObject.NoteList;
import blue.soundObject.NoteParseException;
import electric.xml.Element;

/**
 * <p>
 * Title: blue
 * </p>
 * <p>
 * Description: an object composition environment for csound
 * </p>
 * <p>
 * Copyright: Copyright (c) 2001-2002
 * </p>
 * <p>
 * Company: steven yi music
 * </p>
 * 
 * @author unascribed
 * @version 1.0
 */

@NoteProcessorPlugin(displayName="SwitchProcessor", position = 120)
public class SwitchProcessor implements NoteProcessor, java.io.Serializable {

    int pfield1 = 4;

    int pfield2 = 5;

    @Override
    public String toString() {
        return "[switch]";
    }

    public String getPfield1() {
        return Integer.toString(pfield1);
    }

    public void setPfield1(String pfield1) {
        this.pfield1 = Integer.parseInt(pfield1);
    }

    public String getPfield2() {
        return Integer.toString(pfield2);
    }

    public void setPfield2(String pfield2) {
        this.pfield2 = Integer.parseInt(pfield2);
    }

    public final void processNotes(NoteList in) throws NoteProcessorException {
        Note temp;
        String tempPField;
        int pcount = 0;
        for (int i = 0; i < in.size(); i++) {
            temp = in.get(i);

            // validate necessary pfields are there
            pcount = temp.getPCount();
            if (pfield1 < 1 || pfield1 >= pcount) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.missingPfield"),
                        pfield1);
            }
            if (pfield2 < 1 || pfield2 >= pcount) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.missingPfield"),
                        pfield2);
            }

            tempPField = temp.getPField(pfield1);
            temp.setPField(temp.getPField(pfield2), pfield1);
            temp.setPField(tempPField, pfield2);
        }
    }

    public static void main(String[] args) {
        NoteList n = new NoteList();

        for (int i = 0; i < 10; i++) {
            try {
                n.add(Note.createNote("i1 " + (i * 2) + " 2 3 4"));
            } catch (NoteParseException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        System.out.println("before: \n\n" + n + "\n\n");

        SwitchProcessor sp = new SwitchProcessor();
        sp.setPfield1("5");
        sp.setPfield2("4");
        try {
            sp.processNotes(n);
        } catch (NoteProcessorException ex) {
            System.out.println("Exception: " + ex.getMessage());
        }

        System.out.println("after: \n\n" + n + "\n\n");
    }

    public static NoteProcessor loadFromXML(Element data) {
        SwitchProcessor ap = new SwitchProcessor();

        ap.setPfield1(data.getElement("pfield1").getTextString());
        ap.setPfield2(data.getElement("pfield2").getTextString());

        return ap;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#saveAsXML()
     */
    public Element saveAsXML() {
        Element retVal = new Element("noteProcessor");
        retVal.setAttribute("type", this.getClass().getName());

        retVal.addElement("pfield1").setText(this.getPfield1());
        retVal.addElement("pfield2").setText(this.getPfield2());

        return retVal;
    }
}

File: packages/blue-data/src/note-processors/switch-processor.ts
import { NoteProcessor } from './note-processor';
import { NoteProcessorException } from './note-processor-exception';
import { NoteList } from '../sound-objects/note-list';
import { Element } from '../serialization/xml-reader';

const JAVA_TYPE = 'blue.noteProcessor.SwitchProcessor';

export class SwitchProcessor extends NoteProcessor {
  private _pfield1 = 4;
  private _pfield2 = 5;

  constructor();
  constructor(src: SwitchProcessor);
  constructor(src?: SwitchProcessor) {
    super();
    if (src) {
      this._pfield1 = src._pfield1;
      this._pfield2 = src._pfield2;
    }
  }

  getPfield1(): string {
    return this._pfield1.toString();
  }
  setPfield1(pfield1: string): void {
    this._pfield1 = parseInt(pfield1, 10);
  }

  getPfield2(): string {
    return this._pfield2.toString();
  }
  setPfield2(pfield2: string): void {
    this._pfield2 = parseInt(pfield2, 10);
  }

  override process(notes: NoteList): NoteList {
    for (const note of notes) {
      const pcount = note.getPCount();
      if (this._pfield1 < 1 || this._pfield1 >= pcount) {
        throw new NoteProcessorException('Missing pfield', this._pfield1);
      }
      if (this._pfield2 < 1 || this._pfield2 >= pcount) {
        throw new NoteProcessorException('Missing pfield', this._pfield2);
      }

      const tempPField = note.getPField(this._pfield1);
      note.setPField(note.getPField(this._pfield2)!, this._pfield1);
      note.setPField(tempPField!, this._pfield2);
    }
    return notes;
  }

  override getDisplayName(): string {
    return 'SwitchProcessor';
  }

  override deepCopy(): SwitchProcessor {
    return new SwitchProcessor(this);
  }

  saveAsXML(): Element {
    const elem = new Element('noteProcessor');
    elem.setAttribute('type', JAVA_TYPE);
    elem.addElement('pfield1').setText(this.getPfield1());
    elem.addElement('pfield2').setText(this.getPfield2());
    return elem;
  }

  static loadFromXML(data: Element): SwitchProcessor {
    const proc = new SwitchProcessor();
    const p1 = data.getTextString('pfield1');
    if (p1 !== null) proc._pfield1 = parseInt(p1, 10);
    const p2 = data.getTextString('pfield2');
    if (p2 !== null) proc._pfield2 = parseInt(p2, 10);
    return proc;
  }
}
